bindings/rust/s2n-tls/src/fingerprint.rs

```diff
@@ -0,0 +1,497 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Methods to fingerprint ClientHellos.
+//!
+//! See [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/fingerprint.h).
+
+use crate::{
+    client_hello::ClientHello,
+    error::{Error, Fallible},
+    ffi::*,
+};
+
+use core::ptr::NonNull;
+
+#[non_exhaustive]
+#[derive(Copy, Clone)]
+pub enum FingerprintType {
+    JA3,
+}
+
+const MD5_HASH_SIZE: u32 = 16;
+
+impl From<FingerprintType> for s2n_tls_sys::s2n_fingerprint_type::Type {
+    fn from(value: FingerprintType) -> Self {
+        match value {
+            FingerprintType::JA3 => s2n_tls_sys::s2n_fingerprint_type::FINGERPRINT_JA3,
+        }
+    }
+}
+
+/// A fingerprint operation.
+pub struct Fingerprint<'a>(&'a mut NonNull<s2n_fingerprint>);
+
+impl Fingerprint<'_> {
+    /// Size of the fingerprint hash.
+    /// See s2n_fingerprint_get_hash_size in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/fingerprint.h).
+    pub fn hash_size(&self) -> Result<u32, Error> {
+        let mut hash_size = 0;
+        unsafe { s2n_fingerprint_get_hash_size(self.0.as_ptr(), &mut hash_size).into_result() }?;
+        Ok(hash_size)
+    }
+
+    /// Calculate the fingerprint hash as a hex string.
+    /// See s2n_fingerprint_get_hash in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/fingerprint.h).
+    pub fn hash(&mut self, hash_hex: &mut String) -> Result<u32, Error> {
```

Comment:
So I went that route, and it might have gotten slightly out of hand? `pub fn hash(&mut self, hash_hex: &mut [u8]) -> Result<&str, Error>` worked but maybe looked a little silly, because applications would have to keep the Fingerprint + the [u8] together to avoid scope issues, and they'd definitely keep reusing the same [u8]. It was awkward to call in my tests. So then I thought "why don't they just set the [u8] on the Builder". That got a little messier because I had to Box the [u8].

So I've ended up just allocating the memory for the caller, which does make it very nice and easy to reuse across calls. It should be a two-way door too-- we could always let customers set custom memory on the Builder to override the automatic allocations. I'm skeptical there will be a use case for that though: as long as the Fingerprint is reused for all ClientHellos in a thread, the memory is cheap.